src/anyio/_core/_compat.py

```diff
@@ -36,7 +36,7 @@ async def maybe_async(__obj: 'DeprecatedAwaitable') -> None:
     ...
 
 
-async def maybe_async(__obj: AnyDeprecatedAwaitable) -> Union[TaskInfo, float, list, None]:
+async def maybe_async(__obj: AnyDeprecatedAwaitable) -> Union[TaskInfo, float, List[T], None]:
```

Comment:
Did you mean:
```python
async def maybe_async(__obj: AnyDeprecatedAwaitable[T]) -> Union[TaskInfo, float, List[T], None]:
```
?

Reply:
Same with `_warn_deprecation`:

```python
def _warn_deprecation(awaitable: AnyDeprecatedAwaitable[Any], stacklevel: int = 1) -> None:
```

With exception, that type variable isn't needed here, return type doesn't depend on argument types, `Any` is enough.

Reply:
I've never seen or heard of a `Union` being further parametrized. That makes no sense. Most items in it don't have parameters, and only `DeprecatedAwaitableList` does but I already added the parameter to that one (which fixed the errors).

Reply:
See examples please:

```python
>>> from typing import Union, TypeVar, List
>>>
>>> X = TypeVar('X')
>>> Y = TypeVar('Y')
>>>
>>> U = Union[X, Y]
>>> U[int, str]
typing.Union[int, str]
>>>
>>> U = Union[int, str, List[X]]
>>> U[bool]
typing.Union[int, str, typing.List[bool]]
```

When you use `TypeVar` as type parameter for any generic type, you are only creating another **generic** type which must be further parametrized by concrete type to be _known type_.

Pyright doesn't complain about this because of bug (see https://github.com/agronholm/anyio/pull/397#discussion_r757828005, https://github.com/microsoft/pyright/issues/2625).

Reply:
I found a description of this behavior in [PEP-484 (Type aliases)](https://www.python.org/dev/peps/pep-0484/#type-aliases)

Reply:
Yes, PEP 484 provides for generic type aliases. A type alias is considered generic if it uses one or more type variables within its definition. A generic type alias must be specialized (i.e., type arguments must be supplied) when it is used in a type annotation.

Reply:
I've added the type parameter now, but now PyCharm complains:
![image](https://user-images.githubusercontent.com/130003/143765371-bf0b8de6-75ee-4107-9623-f3ae59dc7228.png)

Reply:
Not only that, but the code no longer runs on **any** version of Python:
`TypeError: typing.Union[_ForwardRef('DeprecatedAwaitable'), _ForwardRef('DeprecatedAwaitableFloat'), _ForwardRef('DeprecatedAwaitableList[T]'), object] is not a generic class`

Reply:
I think you should use forward references to `AnyDeprecatedAwaitable[T]` because of forward references in it's definition.

Reply:
I take it this is because of the forward references? If I make a union with an immediate reference to a parametrized type, it works fine.

Reply:
Okay, all the tests pass now. Do you have any further concerns about the changes or should I merge?

Reply:
Please fix `_warn_deprecation` too.

Reply:
Right, that one. Can I use `Any` there as I don't do anything with the argument?

Reply:
Of course. You can even use `object`.

`Any` allows to do whatever you want with value, `object`does not allow anything (as far as I understand).

Reply:
Doesn't look like `object` works though:
`src/anyio/_core/_compat.py:141: error: Argument 1 to "_warn_deprecation" has incompatible type "DeprecatedAwaitableList[T]"; expected "Union[DeprecatedAwaitable, DeprecatedAwaitableFloat, DeprecatedAwaitableList[object], TaskInfo]"
`

Reply:
I have to leave for a bit now, but I take it this was the last missing piece?

Reply:
Oh, I forgot that `List[T]` is invariant.

Reply:
> I have to leave for a bit now, but I take it this was the last missing piece?

Yeah, thank you!

Reply:
Thanks for the help! I'll merge this now.